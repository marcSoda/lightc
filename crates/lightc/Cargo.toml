[package]
name = "lightc"
version = "0.1.0"
edition = "2021"

[dependencies]
codegen = { path = "../codegen" }
common = { path = "../common" }
lexer = { path = "../lexer" }
parser = { path = "../parser" }
type_checker = { path = "../type_checker" }

clap = { version = "3.0.14", features = ["derive"] }
inkwell = { git = "https://github.com/TheDan64/inkwell", branch = "master", features = ["llvm13-0"] }
llvm-sys = "130"
tempfile = "3.3.0"
