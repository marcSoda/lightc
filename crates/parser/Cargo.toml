[package]
name = "parser"
version = "0.0.0"
edition = "2021"

[dependencies]
serde = { version = "1.0.136", features = ["derive"] }

ast = { path = "../ast" }
common = { path = "../common" }
symbol_table = { path = "../symbol_table" }

[dev-dependencies]
insta = { version = "1.12.0", features = ["glob"] }

lexer = { path = "../lexer" }
