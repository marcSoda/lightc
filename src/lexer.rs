use std::iter::Peekable;
use serde::Serialize;

use crate::token::{Symbol, Token, TokenType};

type LexResult = std::result::Result<Token, LexError>;

pub(crate) struct Lexer {
    stream: Peekable<StreamIter<char>>,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        Lexer {
            stream: StreamIter::new(input).peekable(),
        }
    }

    fn lex(&mut self) -> LexResult {
        let cur = match self.stream.next() {
            Some(cur) => cur,
            None => return Ok(Token::default()),
        };

        if cur.value.is_ascii_whitespace() {
            while let Some(c) = self.stream.peek() {
                if !c.value.is_ascii_whitespace() {
                    return self.lex();
                }
                self.stream.next();
            }
            return self.lex(); // Eat trailing newline
        }

        // Comments
        if cur == '/' && matches!(self.stream.peek(), Some(c) if *c == '/') {
            while let Some(c) = self.stream.next() {
                if c == '\n' {
                    return self.lex();
                }
            }
            return self.lex(); // Eat trailing comment
        }

        // Identifiers: keywords, function names, and variables
        if cur.value.is_ascii_alphabetic() {
            let mut identifier = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '_' {
                    identifier.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            let tt = match identifier.as_str() {
                "else" => TokenType::Else,
                "extern" => TokenType::Extern,
                //"u64" => TokenTypeVarType(TokenTypeU64),
                "fn" => TokenType::Fn,
                "for" => TokenType::For,
                "if" => TokenType::If,
                "let" => TokenType::Let,
                _ => TokenType::Ident(identifier),
            };

            return Ok(Token::from((tt, cur)));
        }

        // Numbers
        if cur.value.is_ascii_digit() {
            let mut num = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '.' {
                    num.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            return match num.parse() {
                Ok(n) => Ok(Token::from((TokenType::Int(n), cur))),
                Err(_) => Err(LexError::from((format!("Invalid number: {}", num), cur))),
            };
        }

        // Logical operators
        if let Some(next) = self.stream.peek() {
            match cur.value {
                '=' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::Eq), cur)));
                }
                '&' if next == &'&' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::And), cur)));
                }
                '|' if next == &'|' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::Or), cur)));
                }
                _ => (),
            }
        }

        // Everything else
        let tt = match cur.value {
            '+' => TokenType::Op(Symbol::Plus),
            '-' => TokenType::Op(Symbol::Minus),
            '*' => TokenType::Op(Symbol::Mult),
            '/' => TokenType::Op(Symbol::Div),
            '^' => TokenType::Op(Symbol::Pow),
            '>' => TokenType::Op(Symbol::Gt),
            '<' => TokenType::Op(Symbol::Lt),
            '!' => TokenType::Op(Symbol::Not),
            '=' => TokenType::Op(Symbol::Assign),
            '}' => TokenType::CloseBrace,
            ')' => TokenType::CloseParen,
            ':' => TokenType::Colon,
            ',' => TokenType::Comma,
            '{' => TokenType::OpenBrace,
            '(' => TokenType::OpenParen,
            ';' => TokenType::Semicolon,
            c => {
                return Err(LexError::from((format!("Unknown character: {}", c), cur)));
            }
        };

        Ok(Token::from((tt, cur)))
    }
}

impl Iterator for Lexer {
    type Item = LexResult;

    fn next(&mut self) -> Option<Self::Item> {
        match self.lex() {
            Ok(Token {
                tt: TokenType::Eof, ..
            }) => None,
            x => Some(x),
        }
    }
}

// Provides additional context for each source character
#[derive(Debug, PartialEq)]
struct ContextPoint<T> {
    value: T,
    line: usize,
    column: usize,
}

impl<T> ContextPoint<T> {
    fn new(value: T, line: usize, column: usize) -> Self {
        ContextPoint {
            value,
            line: line + 1,
            column: column + 1,
        }
    }
}

impl PartialEq<char> for ContextPoint<char> {
    fn eq(&self, other: &char) -> bool {
        self.value == *other
    }
}

impl<T> From<(TokenType, ContextPoint<T>)> for Token {
    fn from((ty, ContextPoint { line, column, .. }): (TokenType, ContextPoint<T>)) -> Self {
        Token {
            tt: ty,
            line,
            column,
        }
    }
}

// Iterator for the source character stream. Supports line and column context.
struct StreamIter<T> {
    lines: Vec<Vec<T>>,
    line: usize,
    column: usize,
}

impl StreamIter<char> {
    fn new(input: &str) -> Self {
        StreamIter {
            lines: input
                .split_inclusive('\n') // can't use .lines() or we lose '\n'
                .map(|l| l.chars().collect())
                .collect(),
            line: 0,
            column: 0,
        }
    }
}

impl Iterator for StreamIter<char> {
    type Item = ContextPoint<char>;

    fn next(&mut self) -> Option<Self::Item> {
        let line = self.lines.get(self.line);
        let cc = line?
            .get(self.column)
            .map(|c| ContextPoint::new(*c, self.line, self.column))
            .or_else(|| {
                self.line += 1;
                self.column = 0;
                self.lines.get(self.line).and_then(|line| {
                    line.get(self.column)
                        .map(|c| ContextPoint::new(*c, self.line, self.column))
                })
            });
        self.column += 1;
        cc
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) struct LexError {
    message: String,
    line: usize,
    column: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Lexing error: {} at {}:{}",
            self.message, self.line, self.column
        )
    }
}

impl std::error::Error for LexError {}

impl<T> From<(String, ContextPoint<T>)> for LexError {
    fn from((msg, cp): (String, ContextPoint<T>)) -> Self {
        LexError {
            message: msg,
            line: cp.line,
            column: cp.column,
        }
    }
}

#[cfg(test)]
mod test {
    use insta::{assert_yaml_snapshot, glob, with_settings};
    use std::fs;

    use super::*;

    #[test]
    fn test_lexer() {
        with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            glob!("tests/inputs/lexer/*.input", |path| {
                let input = fs::read_to_string(path).unwrap();
                assert_yaml_snapshot!(Lexer::new(&input).collect::<Result<Vec<_>, _>>());
            });
        });
    }

    #[test]
    fn test_stream_next() {
        let input = "\
abc
def
ghi
";
        let mut stream = StreamIter::new(input);
        assert_eq!(ContextPoint::new('a', 0, 0), stream.next().unwrap());
        assert_eq!(ContextPoint::new('b', 0, 1), stream.next().unwrap());
        assert_eq!(ContextPoint::new('c', 0, 2), stream.next().unwrap());
        assert_eq!(ContextPoint::new('\n', 0, 3), stream.next().unwrap());
        assert_eq!(ContextPoint::new('d', 1, 0), stream.next().unwrap());
        assert_eq!(ContextPoint::new('e', 1, 1), stream.next().unwrap());
        assert_eq!(ContextPoint::new('f', 1, 2), stream.next().unwrap());
        assert_eq!(ContextPoint::new('\n', 1, 3), stream.next().unwrap());
        assert_eq!(ContextPoint::new('g', 2, 0), stream.next().unwrap());
        assert_eq!(ContextPoint::new('h', 2, 1), stream.next().unwrap());
        assert_eq!(ContextPoint::new('i', 2, 2), stream.next().unwrap());
        assert_eq!(ContextPoint::new('\n', 2, 3), stream.next().unwrap());
        assert_eq!(None, stream.next());
        assert_eq!(None, stream.next());
    }

    #[test]
    fn test_lex() {
        use Symbol::*;
        use TokenType::*;

        let input = "\
let foo = 14
// bar
/ not_a_comment
baz
";
        let mut lexer = Lexer::new(input);
        assert_eq!(Ok(Token::new(Let, 1, 1)), lexer.lex());
        assert_eq!(Ok(Token::new(Ident("foo".to_string()), 1, 5)), lexer.lex());
        assert_eq!(Ok(Token::new(Op(Assign), 1, 9)), lexer.lex());
        assert_eq!(Ok(Token::new(Int(14), 1, 11)), lexer.lex());
        assert_eq!(Ok(Token::new(Op(Div), 3, 1)), lexer.lex());
        assert_eq!(
            Ok(Token::new(Ident("not_a_comment".to_string()), 3, 3)),
            lexer.lex()
        );
        assert_eq!(Ok(Token::new(Ident("baz".to_string()), 4, 1)), lexer.lex());
        assert_eq!(Ok(Token::default()), lexer.lex());
    }
}
