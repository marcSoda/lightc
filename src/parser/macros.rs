#[macro_export]
macro_rules! expect_next_token {
    // Matches patterns like TokenType::Let
    ($ts:expr, $( $e:tt::$v:tt )|+ , $err:expr) => {
        let t = $ts.next();
        match t {
            $( Some(Token { tt: $e::$v, .. }) => (), )+
            _ => {
                return Err(ParseError::from((
                    format!(
                        "{}. Got {}",
                        $err.to_string(),
                        t.map_or(TokenType::Eof, |x| x.tt.clone())
                    ),
                    t.unwrap_or(&Token::default())
                )))
            }
        }
    };

    // Matches patterns like TokenType::Op(Symbol::Assign)
    ($ts:expr, $( $e:tt::$v:tt(Symbol::$s:tt) )|+ , $err:expr) => {
        let t = $ts.next();
        match t {
            $( Some(Token { tt: $e::$v(Symbol::$s), .. }) => (), )+
            _ => {
                return Err(ParseError::from((
                    format!(
                        "{}. Got {}",
                        $err.to_string(),
                        t.map_or(TokenType::Eof, |x| x.tt.clone())
                    ),
                    t.unwrap_or(&Token::default())
                )))
            }
        }
    };

    // Matches patterns like TokenType::Ident(_) and used for return value
    ($ts:expr, $t:tt::$v:tt($_:tt), $err:expr) => {
        {
            let t = $ts.next();
            match t {
                Some(Token { tt: $t::$v(t), .. }) => t,
                _ => {
                    return Err(ParseError::from((
                        format!(
                            "{}. Got {}",
                            $err.to_string(),
                            t.map_or(TokenType::Eof, |x| x.tt.clone())
                        ),
                        t.unwrap_or(&Token::default())
                    )))
                }
            }
        }
    };
}

// Matches token type and executes $and or returns None
#[macro_export]
macro_rules! token_is_and_then {
    ($to:expr, $tt:pat, $and:expr) => {
        match $to {
            Some(Token { tt: $tt, .. }) => Some($and),
            _ => None,
        }
    };
}
