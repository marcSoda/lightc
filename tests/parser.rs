use lightc::lexer::Lexer;
use lightc::parser::*;

fn ast_to_string(ast: &[AstNode]) -> String {
    if ast.len() == 1 {
        return ast[0].to_string();
    }

    let mut s = String::new();
    for node in ast {
        s = s + &node.to_string() + "\n";
    }
    s
}

#[test]
fn test_parser_single_num() {
    let input = "19";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "19";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_two_num_expr() {
    let input = "19 + 21";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(+ 19 21)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_three_num_expr() {
    let input = "19 + 21 + -40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(+ (+ 19 21) (- 40))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_precedence_expr() {
    let input = "19 + 21 * 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(+ 19 (* 21 40))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "19 * 21 - 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(- (* 19 21) 40)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "19 - 21 + 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(+ (- 19 21) 40)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "19 - 21 * 20 + 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(+ (- 19 (* 21 20)) 40)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_right_assoc_expr() {
    let input = "19 ^ 21 ^ 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(^ 19 (^ 21 40))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "19 ^ 21 + 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(+ (^ 19 21) 40)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "19 ^ 21 ^ 40 / 2";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(/ (^ 19 (^ 21 40)) 2)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_multiple_exprs() {
    let input = "19 ^ 21 ^ 40 19 - 21 * 20 + 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(^ 19 (^ 21 40))\n(+ (- 19 (* 21 20)) 40)\n";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_paren_precedence_expr() {
    let input = "(19 + 21) / 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(/ (+ 19 21) 40)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_paren_complex_precedence_expr() {
    let input = "3 * ((19 + 21) - 5) / 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(/ (* 3 (- (+ 19 21) 5)) 40)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_excessive_parens_expr() {
    let input = "(((0)))";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "0";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_simple_ident_expr() {
    let input = "19 + a + 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(+ (+ 19 a) 40)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_nullary_call_expr() {
    let input = "a()";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(a)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_call_expr() {
    let input = "a(b, c)";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(a b c)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_call_with_expr_as_arg() {
    let input = "a(b + 1, c - 2 / 4)";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(a (+ b 1) (- c (/ 2 4)))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_call_with_call_as_arg() {
    let input = "a(b(c))";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(a (b c))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_func_def() {
    let input = "\
fn a(b, c) {
    19 + a + 40
}
";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(define (a b c) (+ (+ 19 a) 40))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_func_def_multi_line() {
    let input = "\
fn a(b, c) {
    19 + a + 40
    b + a
}
";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(define (a b c) (+ (+ 19 a) 40) (+ b a))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_empty_func_def() {
    let input = "fn a() {}";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(define (a))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_gt_lt() {
    let input = "a + b > 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(> (+ a b) 40)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "a < b / 40";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(< a (/ b 40))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_extern() {
    let input = "extern cos(x)";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(define (cos x))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_if_only() {
    let input = "if a > b { foo }";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    println!("{:?}", tokens);
    let parser = Parser::new(&tokens);
    let ast = "(if (> a b) foo)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_if_else() {
    let input = "if a > b { foo } else { bar }";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(if (> a b) foo bar)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "if a > b { foo } else if a < b { bar } else { baz }";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(if (> a b) foo (if (< a b) bar baz))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_for() {
    let input = "for let x = 1; x < 10; 1 { foo }";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    println!("{:?}", tokens);
    let parser = Parser::new(&tokens);
    let ast = "(for (let x 1) (< x 10) 1 foo)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_unary() {
    let input = "-21";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(- 21)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "-a * 2";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(* (- a) 2)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "3 - -21";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(- 3 (- 21))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "--21";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(- (- 21))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}

#[test]
fn test_parser_logical_ops() {
    let input = "x == 1";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(== x 1)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "x && 1";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(&& x 1)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "x || 1";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(|| x 1)";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "x == 1 && y == 2 || z == 3";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(|| (&& (== x 1) (== y 2)) (== z 3))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);

    let input = "x == 1 && (y == 2 || z == 3)";
    let tokens = Lexer::new(input).collect::<Result<Vec<_>, _>>().unwrap();
    let parser = Parser::new(&tokens);
    let ast = "(&& (== x 1) (|| (== y 2) (== z 3)))";
    assert_eq!(ast_to_string(&parser.parse().unwrap()), ast);
}
