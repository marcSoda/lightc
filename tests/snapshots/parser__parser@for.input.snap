---
source: tests/parser.rs
assertion_line: 36
expression: asts
input_file: tests/inputs/parser/for.input

---
- - Ok:
      - Expr:
          For:
            var_name: x
            start:
              Num:
                value: 1
            cond:
              BinOp:
                sym: Lt
                lhs:
                  Var:
                    name: x
                rhs:
                  Num:
                    value: 10
            step:
              Num:
                value: 1
            body:
              - Var:
                  name: foo
  - (for (let x 1) (< x 10) 1 foo)
- - Err: "Expecting 'let' after for"
  - "Expecting 'let' after for"
- - Err: Expecting identifier after let
  - Expecting identifier after let
- - Err: "Expecting '=' after identifer"
  - "Expecting '=' after identifer"
- - Err: Invalid expression
  - Invalid expression
- - Err: "Expecting ';' after start"
  - "Expecting ';' after start"

